crates/core/app/src/app_version/component.rs

```diff
@@ -0,0 +1,138 @@
+use std::fmt::Write as _;
+
+use anyhow::{anyhow, Context};
+use cnidarium::{StateDelta, StateRead, StateWrite, Storage};
+
+use super::APP_VERSION;
+
+fn version_to_software_version(version: u64) -> &'static str {
+    match version {
+        1 => "v0.70.x",
+        2 => "v0.73.x",
+        3 => "v0.74.x",
+        4 => "v0.75.x",
+        5 => "v0.76.x",
+        6 => "v0.77.x",
+        7 => "v0.79.x",
+        8 => "v0.80.x",
+        _ => "unknown",
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+enum CheckContext {
+    Running,
+    Migration,
+}
+
+fn check_version(ctx: CheckContext, expected: u64, found: Option<u64>) -> anyhow::Result<()> {
+    let found = match (expected, found) {
+        (x, Some(y)) if x != y => y,
+        _ => return Ok(()),
+    };
+    match ctx {
+        CheckContext::Running => {
+            let expected_name = version_to_software_version(expected);
+            let found_name = version_to_software_version(expected);
+            let mut error = String::new();
+            error.push_str("app version mismatch:\n");
+            write!(
+                &mut error,
+                "  expected {} (penumbra {})\n",
+                expected, expected_name
+            )?;
+            write!(&mut error, "  found {} (penumbra {})\n", found, found_name)?;
+            write!(
+                &mut error,
+                "make sure you're running penumbra {}",
+                expected_name
+            )?;
+            Err(anyhow!(error))
+        }
+        CheckContext::Migration => {
+            let expected_name = version_to_software_version(expected);
+            let found_name = version_to_software_version(expected);
+            let mut error = String::new();
+            error.push_str("app version mismatch:\n");
+            write!(
+                &mut error,
+                "  expected {} (penumbra {})\n",
+                expected, expected_name
+            )?;
+            write!(&mut error, "  found {} (penumbra {})\n", found, found_name)?;
+            write!(
+                &mut error,
+                "this migration should be run with penumbra {} instead",
+                version_to_software_version(expected + 1)
+            )?;
+            Err(anyhow!(error))
+        }
+    }
+}
+
+fn state_key() -> Vec<u8> {
+    b"penumbra_app_version_safeguard".to_vec()
```

Comment:
Can we move this to the state key module with the rest of the app level state keys?